librz/analysis/il_trace.c

```diff
@@ -0,0 +1,208 @@
+// SPDX-FileCopyrightText: 2021 Heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+/**
+ * \file il_trace.c
+ * \brief new rizin IL trace implementation
+ *
+ * provide operations to new IL trace structure to record
+ * the memory changes and register changes.
+ * TODO : Should be moved to librz/il after integrations with new il
+ *      : should move the prototypes and trace structure to new header, too
+ * prototypes in <rz_analysis.h>
+ * Used by : analysis_tp.c, debug/trace.c
+**/
+
+#include <rz_util.h>
+#include <rz_analysis.h>
+
+static void free_reg_op(RzILTraceRegOp *reg_op) {
+	if (reg_op) {
+		if (reg_op->reg_name) {
+			free(reg_op->reg_name);
+		}
```

Comment:
it is never necessary to do `if (x) free (x)`. If it is necessary, it means the `free` function needs to gracefully handle the case when the arg is NULL.